### ISSUE TITLE
How do you deal with multiple files ?

### ISSUE THREAD

**Issue by Iuliean** [CrowCpp/Crow#185]
My app has a class called site in a site.h headder with the site.cpp file and i wish to include that in my main.cpp but if i define the `#define CROW_MAIN` in the site.h i get and error that ```/usr/bin/ld: main.o:(.bss+0x0): multiple definition of `crow::mime_types[abi:cxx11]'; site.o:(.bss+0x0): first defined here
/usr/bin/ld: main.o:(.data+0x0): multiple definition of `crow::detail::dumb_timer_queue::tick'; site.o:(.data+0x0): first defined here``` but if i put the define in only one of them it complains in the other that the define does not exitst. Sorry i posted the issue empty by mistake

**Comment by The-EDev** [comment]
Could you please elaborate on what you mean by multiple files?

**Comment by The-EDev** [comment]
Normally you're only supposed to have `#define CROW_MAIN` defined in your main .cpp file **above** any `#include`s. Which part are you including the Crow headers in?

**Comment by Iuliean** [comment]
I'm including it in `site.h` at the top just right after `#pragma once` then `site.h` get's included in `site.cpp` and `main.cpp`. i need it in the headder since some funcs require `const crow::request` as parameters

**Comment by The-EDev** [comment]
`#define CROW_MAIN` should only be defined in your `main.cpp` (I'm guessing that's where you're creating and running your Crow app)

**Comment by Iuliean** [comment]
well yes but when it compile's `site.cpp` it complains that ```In file included from site.h:2,
                 from site.cpp:1:
dependencies/includes/crow/crow_all.h:39:25: error: declaration of ‘constexpr’ variable ‘crow::VERSION’ is not a definition
   39 |   extern constexpr char VERSION[];``` i can post the full code if that's easier

**Comment by The-EDev** [comment]
> i can post the full code if that's easier

That would make things a lot more convenient. But please only do it if you're comfortable making your code public. otherwise you can create a private repository and point me to it.

**Comment by Iuliean** [comment]
No it's ok it's just some little project i am working on.
```cpp
//site.cpp
#include "site.h"
#include "json/json.hpp"

Site::Site(Controller* otherController)
{
    this->cntrl = otherController;
    nlohmann::json j = configFile.read();

    this->port = j["site"]["port"];
    this->password = j["site"]["password"];


    
    
    CROW_ROUTE(this->app, "/").methods(crow::HTTPMethod::GET)([this](){return this->loginPageGet();});
    CROW_ROUTE(this->app, "/").methods(crow::HTTPMethod::POST)([this](const crow::request& req)
                                                        {return this->loginPagePost(req);});
    CROW_ROUTE(this->app, "/settings").methods(crow::HTTPMethod::POST)([this](const crow::request& req)
                                                        {return this->settingsPage(req);});

}

void Site::run()
{
    app.port(this->port).multithreaded().run();
}

//Private

std::string Site::loginPageGet()
{
    CROW_LOG_INFO << "Serving login page";
    auto page = crow::mustache::load("login.html");
    return page.render();
}

std::string Site::loginPagePost(const crow::request& req)
{
    CROW_LOG_INFO << "Login post request";

    crow::multipart::message passwd(req);

    return passwd.parts[0].body == this->password ? this->dashboardGet() : "Wrong password";
}

std::string Site::dashboardGet()
{
    CROW_LOG_INFO << "Serving Dashboard page";

    crow::mustache::context ctx;

    ctx["threshold"]    = this->cntrl->getThreshold();
    ctx["temp"]         = this->cntrl->getTemp();
    ctx["range"]        = this->cntrl->getRange();
    ctx["state"]        = this->cntrl->getState() ? "On" : "Off";

    auto page = crow::mustache::load("dashboard.html");

    return page.render(ctx);
}
std::string Site::settingsPage(const crow::request& req)
{
    CROW_LOG_INFO << "Setting page call";

    crow::multipart::message msg(req);

    this->cntrl->setThreshold(std::stof(msg.parts[0].body));
    this->cntrl->setRange(std::stof(msg.parts[1].body));

    CROW_LOG_INFO << "New threshold: " << this->cntrl->getThreshold() << 
                    ". New range: " << this->cntrl->getRange();
    
    return this->dashboardGet();
}
```
```cpp
//site.h
#pragma once
#include "crow/crow_all.h"
#include "file.h"
#include "controller.h"

#include <string>

class Site
{
private:
    File configFile{"config.json"};
    Controller* cntrl;

    crow::SimpleApp app;

    int port;
    std::string password;

public:
    Site(Controller* otherController);

    void run();

private:
    std::string loginPageGet();
    std::string loginPagePost(const crow::request& req);
    
    std::string dashboardGet();
    std::string settingsPage(const crow::request& req);

};
```
```cpp
//main.cpp
#define CROW_MAIN
#include "site.h"
#include "json/json.hpp"
#include "controller.h"

#include <fstream>


int main()
{   
    //init();

    Controller ctrl;
    Site site(&ctrl);
    site.run();

    //ctrl.print();

    //ctrl.setRange(40);

    //ctrl.print();

    ctrl.run();


    //app.loglevel(crow::LogLevel::Info);

    
    //CROW_ROUTE(app, "/").methods(crow::HTTPMethod::POST)(login_page_post);
    //CROW_ROUTE(app, "/settings").methods(crow::HTTPMethod::POST)(settings_page);

    //app.port(18080).multithreaded().run();
    return 0;
}
```

**Comment by The-EDev** [comment]
I removed anything unrelated to these 3 files and it seems to compile fine. Are you running `master` or one of the releases? (I tested with `v0.3+2`)

**Comment by Iuliean** [comment]
I actually don't remember. i'll switch the headder with the release one now and i'll get back to you. I am compiling with `-lpthread -wiringPi` and i added the boost lib myself to the proj

**Comment by The-EDev** [comment]
NO problem, thanks for letting me know about this problem, I'll investigate it further on `master` branch.

**Comment by The-EDev** [comment]
@Iuliean This turned out to be a legitimate bug, it's been fixed in #186. You shouldn't face this problem in `master` once the PR is merged. Sorry for the inconvenience and thanks again for bringing it up.
